ivi-input-modules/ivi-input-controller/src/ivi-input-controller.c

```diff
@@ -890,33 +916,45 @@ handle_seat_updated_caps(struct wl_listener *listener, void *data)
 }
 
 static void
-handle_seat_destroy(struct wl_listener *listener, void *data)
+destroy_seat(struct seat_ctx *ctx_seat)
 {
-    struct seat_ctx *ctx = wl_container_of(listener, ctx, destroy_listener);
-    struct weston_seat *seat = data;
-    struct wl_resource *resource;
-    struct input_context* input_ctx = ctx->input_ctx;
     struct ivisurface *surf;
-
-    if (ctx->keyboard_grab.keyboard)
-        keyboard_grab_cancel(&ctx->keyboard_grab);
-    if (ctx->pointer_grab.pointer)
-        pointer_grab_cancel(&ctx->pointer_grab);
-    if (ctx->touch_grab.touch)
-        touch_grab_cancel(&ctx->touch_grab);
+    struct wl_resource *resource;
+    if (ctx_seat->keyboard_grab.keyboard) {
+    	keyboard_grab_cancel(&ctx_seat->keyboard_grab);
+        weston_keyboard_end_grab(ctx_seat->keyboard_grab.keyboard);
```

Comment:
only calling weston_keyboard_end_grab should be fine.

Reply:
Based on previous comments and handling of default grab, keyboard focus need to be set to NULL and grab needs to be cancelled. End grab just sets the grab interface to default grab and does not do clean up. It can happen that when drm backend is destroyed leaked things can affect.

Here we are just taking the focus away in grab_cancel and keeping things clean before ending the grab.